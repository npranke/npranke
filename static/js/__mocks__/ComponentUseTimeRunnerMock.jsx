import PropTypes from 'prop-types'
import React from 'react'

import useTimeRunner from '@components/hooks/TimeRunner'

function ComponentUseTimeRunnerMock(props) {
    const { centiseconds, seconds, minutes } = useTimeRunner(
        props.isTimeRunning,
    )

    return (
        <div className="component-use-time-runner-mock">
            <div className="centiseconds">{ centiseconds }</div>
            <div className="seconds">{ seconds }</div>
            <div className="minutes">{ minutes }</div>
        </div>
    )
}

ComponentUseTimeRunnerMock.defaultProps = {
    isTimeRunning: false,
}

ComponentUseTimeRunnerMock.propTypes = {
    isTimeRunning: PropTypes.bool,
}

export default ComponentUseTimeRunnerMock
