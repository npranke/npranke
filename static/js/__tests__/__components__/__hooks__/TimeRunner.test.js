import { act, render, screen } from '@testing-library/react'

import ComponentUseTimeRunnerMock from '@__mocks__/ComponentUseTimeRunnerMock'

jest.useFakeTimers()

describe('TimeRunner', () => {
    beforeEach(() => {
        jest.setSystemTime(0)
    })

    test('return time 00 when isRunning false', () => {
        render(
            <ComponentUseTimeRunnerMock
                isRunning={ false }
                shouldReset={ false }
            />,
        )

        const times = screen.getAllByRole('timer')

        act(() => {
            jest.runOnlyPendingTimers()
            jest.advanceTimersByTime(100)
        })

        expect(
            times[0].textContent,
        ).toEqual('00')
        expect(
            times[1].textContent,
        ).toEqual('00')
        expect(
            times[2].textContent,
        ).toEqual('00')
    })

    test('return time 00 when shouldReset true', () => {
        render(
            <ComponentUseTimeRunnerMock
                isRunning
                shouldReset
            />,
        )

        const times = screen.getAllByRole('timer')

        act(() => {
            jest.runOnlyPendingTimers()
            jest.advanceTimersByTime(100)
        })

        expect(
            times[0].textContent,
        ).toEqual('00')
        expect(
            times[1].textContent,
        ).toEqual('00')
        expect(
            times[2].textContent,
        ).toEqual('00')
    })

    test('formats centiseconds when less than 10', () => {
        render(
            <ComponentUseTimeRunnerMock
                isRunning
                shouldReset={ false }
            />,
        )

        const times = screen.getAllByRole('timer')

        act(() => {
            jest.runOnlyPendingTimers()
            jest.advanceTimersByTime(60)
        })

        expect(
            times[0].textContent,
        ).toEqual('00')
        expect(
            times[1].textContent,
        ).toEqual('00')
        expect(
            times[2].textContent,
        ).toEqual('05')
    })

    test('formats seconds when less than 10', () => {
        render(
            <ComponentUseTimeRunnerMock
                isRunning
                shouldReset={ false }
            />,
        )

        const times = screen.getAllByRole('timer')

        act(() => {
            jest.runOnlyPendingTimers()
            jest.advanceTimersByTime(8020)
        })

        expect(
            times[0].textContent,
        ).toEqual('00')
        expect(
            times[1].textContent,
        ).toEqual('08')
        expect(
            times[2].textContent,
        ).toEqual('01')
    })

    test('formats minutes when less than 10', () => {
        render(
            <ComponentUseTimeRunnerMock
                isRunning
                shouldReset={ false }
            />,
        )

        const times = screen.getAllByRole('timer')

        act(() => {
            jest.runOnlyPendingTimers()
            jest.advanceTimersByTime(124030)
        })

        expect(
            times[0].textContent,
        ).toEqual('02')
        expect(
            times[1].textContent,
        ).toEqual('04')
        expect(
            times[2].textContent,
        ).toEqual('02')
    })

    test('formats centiseconds when over 100', () => {
        render(
            <ComponentUseTimeRunnerMock
                isRunning
                shouldReset={ false }
            />,
        )

        const times = screen.getAllByRole('timer')

        act(() => {
            jest.runOnlyPendingTimers()
            jest.advanceTimersByTime(1050)
        })

        expect(
            times[0].textContent,
        ).toEqual('00')
        expect(
            times[1].textContent,
        ).toEqual('01')
        expect(
            times[2].textContent,
        ).toEqual('04')
    })

    test('formats seconds when over 60', () => {
        render(
            <ComponentUseTimeRunnerMock
                isRunning
                shouldReset={ false }
            />,
        )

        const times = screen.getAllByRole('timer')

        act(() => {
            jest.runOnlyPendingTimers()
            jest.advanceTimersByTime(63070)
        })

        expect(
            times[0].textContent,
        ).toEqual('01')
        expect(
            times[1].textContent,
        ).toEqual('03')
        expect(
            times[2].textContent,
        ).toEqual('06')
    })

    describe('useEffect dependency changes', () => {
        beforeAll(() => {
            jest.spyOn(window, 'setInterval')
        })

        beforeEach(() => {
            jest.clearAllMocks()
        })

        afterAll(() => {
            jest.restoreAllMocks()
        })

        test('setInterval called again with isRunning change', () => {
            const { rerender } = render(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(1)

            rerender(
                <ComponentUseTimeRunnerMock
                    isRunning
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(2)

            rerender(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(3)
        })

        test('setInterval not called again without isRunning change', () => {
            const { rerender } = render(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(1)

            rerender(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(1)
        })

        test('setInterval called again with shouldReset change', () => {
            const { rerender } = render(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(1)

            rerender(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(2)

            rerender(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(3)
        })

        test('setInterval not called again without shouldReset change', () => {
            const { rerender } = render(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(1)

            rerender(
                <ComponentUseTimeRunnerMock
                    isRunning={ false }
                    shouldReset={ false }
                />,
            )

            expect(
                window.setInterval,
            ).toHaveBeenCalledTimes(1)
        })
    })
})

describe('TimeRunner snapshot', () => {
    test('matches snapshot without time runing', () => {
        const { asFragment } = render(
            <ComponentUseTimeRunnerMock
                isRunning={ false }
                shouldReset={ false }
            />,
        )

        expect(asFragment()).toMatchSnapshot()
    })

    test('matches snapshot with time run to 02:36:19', () => {
        const { asFragment, rerender } = render(
            <ComponentUseTimeRunnerMock
                isRunning
                shouldReset={ false }
            />,
        )

        act(() => {
            jest.setSystemTime(0)
            jest.runOnlyPendingTimers()
            jest.advanceTimersByTime(156200)
        })

        rerender(<ComponentUseTimeRunnerMock />)

        expect(asFragment()).toMatchSnapshot()
    })
})
