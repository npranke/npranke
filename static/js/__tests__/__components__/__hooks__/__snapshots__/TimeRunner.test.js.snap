// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeRunner snapshot matches snapshot 1`] = `
<ComponentUseTimeRunnerMock
  isTimeRunning={false}
>
  <div
    className="component-use-time-runner-mock"
  >
    <div
      className="centiseconds"
    >
      00
    </div>
    <div
      className="seconds"
    >
      00
    </div>
    <div
      className="minutes"
    >
      00
    </div>
  </div>
</ComponentUseTimeRunnerMock>
`;
